docs/proposals/20220221-runtime-SDK.md

```diff
@@ -0,0 +1,802 @@
+  implement a strongly typed func that contains the actual extension implementation.
+
+E.g.
+_Note: this is pseudo code, meant to demonstrate that implementing a RuntimeExtension requires only minimal scaffolding;
+the exact details will be defined during implementation, possibly taking advantage of Golang generics._
+
+```go
+var c = catalog.NewCatalog()
+
+func init() {
+    v1alpha2.AddToCatalog(c)
+}
+
+func main() {
+    ctx := context.Background()
+
+    listener, err := net.Listen("tcp", net.JoinHostPort("127.0.0.1", "8082"))
+    if err != nil {
+        panic(err)
+    }
+
+    // Build an HTTP handler for a given operation, calling a strongly typed func at each request
+    beforeUgradeHandler, err := catalogHTTP.NewHandlerBuilder().
+        WithCatalog(c).
+        AddService(&v1alpha2.BeforeUgradeHook{}, doBeforeUpgrade).
+        Build()
+    if err != nil {
+        panic(err)
+    }
+
+    srv := &http.Server{
+        Handler: BeforeUgradeHandler,
+    }
+    if err := srv.Serve(listener); err != nil && err != http.ErrServerClosed {
+        panic(err)
+    }
+}
+
+func doBeforeUpgrade(in *v1alpha2.BeforeUpgradeInput, out *v1alpha2.BeforeUpgradeOutput) error {
+    // your actual implementation...
+    return nil
+}
+```
+
+While writing the actual code of the Runtime Extension a set of important guidelines must be considered:
+
+#### Timeouts
+Runtime Extension processing adds to network request latency, they should run as quickly as possible
+(typically in milliseconds); Cluster Administrator will be allowed to configure how long the Cluster API Runtime
+should wait for a Runtime Extension to respond before treating the call as a failure (max 10s).
+
+####  Availability
+Runtime Extension failure could result in errors in handling Workload’s Clusters lifecycle, and so the implementation
+should be robust, have proper error handling, avoid panics, etc.; Cluster Administrator will be allowed to set up
+failure policies preventing a Runtime Extension failure to have negative effects on the Cluster API Runtime, but
+this option can’t be used in all the use cases.
+
+#### Side Effects
+It is recommended that Runtime Extensions should avoid side effects if possible, which means to operate only on
+the content of the Input/Output sent to them, and not make out-of-band changes.
+If side effects are required, following rules apply:
+
+#### Idempotence
+An idempotent Runtime Extension is able to successfully accomplish its task also in case it has already been completed
+(the Runtime Extension checks current state and changes it only if necessary).
+
+#### Avoid dependencies
+Each Runtime Extension should accomplish its task without dependency or relations with other Runtime Extensions.
+Introducing dependencies across Runtime Extensions makes the system fragile, and it is probably a consequence of
+poor “Separation of Concerns” while designing such components.
+
```

Comment:
Q: Should we mention determinism here too?

I think we don't have to, but for some RuntimeExtensions it might be super important that the same input always results in the same output.

Reply:
Sounds good!